polygeist/tools/cgeist/Lib/CodeGenTypes.cc

```diff
@@ -78,6 +234,259 @@ const CodeGenOptions &CodeGenTypes::getCodeGenOpts() const {
+    if (RetAI.getInAllocaSRet()) {
+      // sret things on win32 aren't void, they return the sret pointer.
+      QualType Ret = FI.getReturnType();
+      mlir::Type Ty = getMLIRType(Ret);
+      unsigned AddressSpace = CGM.getContext().getTargetAddressSpace(Ret);
+      ResultType = getPointerOrMemRefType(Ty, AddressSpace);
+      ResultType = mlir::MemRefType::get(-1, Ty, {}, AddressSpace);
+    } else {
+      ResultType = Builder.getNoneType();
+    }
+    break;
+
+  case clang::CodeGen::ABIArgInfo::Indirect:
+    LLVM_DEBUG(llvm::dbgs() << "RetInfo: ABIArgInfo::Indirect\n");
+    if (!AllowSRet) {
+      // HACK: remove once we can handle function returning a struct.
+      QualType Ret = FI.getReturnType();
+      ResultType = getMLIRType(Ret);
+      break;
+    }
+  case clang::CodeGen::ABIArgInfo::Ignore:
+    LLVM_DEBUG(llvm::dbgs() << "RetInfo: ABIArgInfo::Ignore\n");
+    ResultType = Builder.getNoneType();
+    break;
+
+  case clang::CodeGen::ABIArgInfo::CoerceAndExpand:
+    LLVM_DEBUG(llvm::dbgs() << "RefInfo: ABIArgInfo::CoerceAndExpand\n");
+    llvm_unreachable("not implemented");
+    break;
+  }
+
+  //
+  // Compute the types of the function parameters.
+  //
+  const unsigned NumArgs = IsArrayReturn ? IRFunctionArgs.totalIRArgs() + 1
+                                         : IRFunctionArgs.totalIRArgs();
+  SmallVector<mlir::Type, 8> ArgTypes(NumArgs);
+  LLVM_DEBUG(llvm::dbgs() << "NumArgs = " << NumArgs << "\n");
+
+  // Add type for sret argument.
+  if (AllowSRet && IRFunctionArgs.hasSRetArg()) {
+    llvm_unreachable("not implemented");
+    QualType Ret = FI.getReturnType();
+    mlir::Type Ty = getMLIRType(Ret);
+    unsigned AddressSpace = CGM.getContext().getTargetAddressSpace(Ret);
+    ArgTypes[IRFunctionArgs.getSRetArgNo()] =
+        mlir::MemRefType::get(-1, Ty, {}, AddressSpace);
+  }
+
+  // Add type for inalloca argument.
+  if (AllowInAllocaRet && IRFunctionArgs.hasInallocaArg()) {
+    llvm_unreachable("not implemented");
+    auto ArgStruct = FI.getArgStruct();
+    assert(ArgStruct);
+    //  auto Ty = LLVM::LLVMStructType::getLiteral(TheModule->getContext(),
+    //  ArgTys);
+    // ArgTypes[IRFunctionArgs.getInallocaArgNo()] =
+    //  mlir::MemRefType::get(-1, Ty, {}, 0);
+  }
+
+  // Add in all of the required arguments.
+  unsigned ArgNo = 0;
+  clang::CodeGen::CGFunctionInfo::const_arg_iterator
+      it = FI.arg_begin(),
+      ie = it + FI.getNumRequiredArgs();
+  for (; it != ie; ++it, ++ArgNo) {
+    // Note: 'ArgTy' is the type of the parameter after it as been decayed to
+    // abide by the ABI rules.
+    const QualType &ArgTy = it->type;
+    const clang::CodeGen::ABIArgInfo &ArgInfo = it->info;
+
+    // TODO: Currently cgeist does not handle inserting paddings. Need to
+    // revisit.
+    bool InsertPadding = false;
+
+    // Insert a padding type to ensure proper alignment.
+    if (InsertPadding && IRFunctionArgs.hasPaddingArg(ArgNo)) {
```

Comment:
We have an `llvm_unreachable("not implemented");` currently so when we hit this condition, we will know we need to fix the problem.

Reply:
Yes, I meant in case `!InsertPadding && IRFunctionArgs.hasPaddingArg(ArgNo)`. This is just a suggestion, if you think this would be too verbose, we can skip (this applies to other similar comments).